Added the framework for compiling and running musl app

Azure-stars opened the pull request:
`apps/` folder stores code for all kinds of testcases. 
For some complex applications, it can be compiled elsewhere. Its binaries can be stored in the `apps/` folder directly or via a sub-repository.

Azure-stars committed:
[app] Add testcases for musl libc

Azure-stars committed:
[feat] Support for loading musl app and entering main function

Azure-stars committed:
[syscall] Add some necessary syscalls for base musl applications

Azure-stars committed:
[ci] Add deploy CI and update test CI

equation314 commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,34 @@
+      contents: write
+    env:
+      default-branch: ${{ format('refs/heads/{0}', github.event.repository.default_branch) }}
+    steps:
```
Add `RUSTDOCFLAGS` env:

```
-D rustdoc::broken_intra_doc_links -D missing-docs
```

Azure-stars replied:
Add `RUSTDOCFLAGS` env in Makefile and export it when running cargo doc command

equation314 commented on Cargo.toml:
```diff
@@ -10,11 +10,14 @@ repository = "https://github.com/arceos-org/starry-next"
 axstd = { git = "https://github.com/arceos-org/arceos.git", features = ["paging"] }
 arceos_posix_api = { git = "https://github.com/arceos-org/arceos.git" }
+elf_parser = { git = "https://github.com/Starry-OS/elf-parser.git" }
+numeric-enum-macro = { git = "https://github.com/mexus/numeric-enum-macro" }
```
Can we use the crate in crates.io?

```
numeric-enum-macro = "0.2"
```

equation314 replied:
The [num_enum](https://crates.io/crates/num_enum) crate is more commonly used.

Azure-stars replied:
The `elf_parser` crate has not yet been uploaded to crates.io

equation314 commented on apps/libc/build/x86_64/helloworld_c:
Remove these binaries.

equation314 commented on apps/libc/rust/helloworld/.gitignore:
We already have `.gitignore` in the root directory and this file is no longer needed.

equation314 commented on apps/libc/rust/task/sleep/.gitignore:
Remove this file.

equation314 commented on src/mm.rs:
```diff
@@ -1,3 +1,6 @@
+use alloc::string::ToString;
+use alloc::vec;
+use alloc::vec::Vec;
```
I recommend this style with less number of lines:

```rs
use alloc::{string::ToString, vec, vec::Vec};
```

equation314 commented on src/syscall.rs:
Split this file into multiple files according to the syscall categories, such as modules in `arceos_api`.

equation314 commented on src/syscall.rs:
```diff
@@ -1,21 +1,181 @@
+    Ok(start_addr)
+}
+
+fn sys_mmap(addr: *mut usize, length: usize, prot: i32, flags: i32) -> isize {
```
`fd` and `offset`?

equation314 commented on src/syscall.rs:
```diff
@@ -1,21 +1,181 @@
+            }
+            0
+        }
+        Ok(ArchPrctlCode::ArchSetGs) | Ok(ArchPrctlCode::ArchGetGs) => todo!(),
```
We can get user `GSBASE` via the `IA32_KERNEL_GS_BASE` MSR.

Azure-stars replied:
Do I need to add functions to access gs to HAL（Although the GetGS flags is only avaliable for x86_64）？

equation314 commented on src/syscall.rs:
```diff
@@ -1,21 +1,181 @@
+        }
+        Ok(ArchPrctlCode::ArchGetFs) => {
+            unsafe {
+                *addr = *(axhal::arch::read_thread_pointer() as *mut usize);
```
Why dereference fsbase?

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +76,41 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
+            .unwrap_or_else(|e| panic!("Mapping failed for stack: {:#x?} error: {:?}", vaddr, e));
+        unsafe {
+            core::ptr::copy_nonoverlapping(
+                stack_data.as_ptr().add(idx * memory_addr::PAGE_SIZE_4K),
```
Add `use memory_addr::PAGE_SIZE_4K` to avoid  writing `memory_addr::` many times.

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +76,41 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
+        ustack_top, ustack_base
     );
+    // FIXME: Add more arguments and environment variables
+    let args = vec![app_name.to_string()];
```
Try to make the type of the `args` parameter in `get_app_stack_region` a slice, to avoid heap allocation here.

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +76,41 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
+    // FIXME: Add more arguments and environment variables
+    let args = vec![app_name.to_string()];
+    let (stack_data, ustack_bottom) =
+        elf_parser::get_app_stack_region(args, &Vec::new(), elf_info.auxv, ustack_top, ustack_size);
```
`&Vec::new()` -> `&[]`

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +76,41 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
-    info!("New user address space: {:#x?}", uspace);
-    Ok((elf_info.entry, ustack_top, uspace))
+
+    for (idx, vaddr) in memory_addr::PageIter4K::new(ustack_top, ustack_base)
```
Why we need to iterate the whole stack? `env` and `aux` take up only a small portion of the user stack.

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +71,52 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
 
-    let ustack_top = uspace.end();
-    let ustack_vaddr = ustack_top - crate::USER_STACK_SIZE;
+    let ustack_base = uspace.end();
```
`ustack_end` is more readable

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +71,52 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
-    let ustack_vaddr = ustack_top - crate::USER_STACK_SIZE;
+    let ustack_base = uspace.end();
+    let ustack_size = crate::USER_STACK_SIZE;
+    let ustack_top = ustack_base - ustack_size;
```
Swap the naming of `ustack_top` and `ustack_bottom`, top should be greater than bottom.

equation314 commented on src/mm.rs:
```diff
@@ -73,20 +71,52 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
+        let ustack_bottom_align = VirtAddr::from_usize(ustack_bottom).align_down_4k();
+        let stack_data_offset = ustack_bottom_align - ustack_top;
+        // Only copy data which contains args, envs and auxv.
+        for (idx, vaddr) in PageIter4K::new(ustack_bottom_align, ustack_base)
```
We can add `read` and `write` methods in `axmm::AddrSpace` to copy data.

Azure-stars committed:
[syscall] Split syscalls into multiple files according to the syscall categories

Azure-stars committed:
[build] Add configuration for user space

Azure-stars committed:
[loader] Move IO methods to axmem module and simplify codes.

Azure-stars committed:
[CI] Continue on error for clippy and fmt test

Azure-stars committed:
[CI] Update dependencies for qemu build

hky1999 reviewed (approved):
Never mind, I submit a approve by mistake

hky1999 reviewed (changes requested):
pending

hky1999 commented on apps/libc/c/helloworld/helloworld.c:
We have moved out the `/apps/` directory from ArceOS in https://github.com/arceos-org/arceos/commit/5f67fdaaa8976d4a037b91dd0fbb2b8b0da5f603, should we reintroduce the `/apps/` here?

equation314 replied:
They are temporary. Since our current syscall support is incomplete, it does not work well with arceos-app.

Azure-stars replied:
`/apps/` here is a temporary solution, which stores source codes and binary files for testing.

When the filesystem is supported, these testcases may be moved to a seperate repo.

equation314 commented on configs/aarch64.toml:
```diff
@@ -0,0 +1,17 @@
+# The base address of the user space.
+user_space_base = 0x1000
```
Recommend this style in toml: `user-space-base`

Azure-stars replied:
ok

equation314 commented on src/syscall_imp/fs/io.rs:
```diff
@@ -0,0 +1,14 @@
+pub(crate) fn sys_read(fd: i32, buf: *mut c_void, count: usize) -> isize {
+    api::sys_read(fd, buf, count)
+}
+pub(crate) fn sys_write(fd: i32, buf: *const c_void, count: usize) -> isize {
```
Add blank line

equation314 commented on src/mm.rs:
```diff
@@ -26,67 +34,39 @@ pub fn load_user_app(app_name: &str) -> AxResult<(VirtAddr, VirtAddr, AddrSpace)
-    let ustack_vaddr = ustack_top - crate::USER_STACK_SIZE;
+    // The user stack is divided into two parts:
+    // `ustack_bottom` -> `ustack top`: It is the stack space that users actually read and write.
+    // `ustack_top` -> `ustack end`: It is the space that contains the arguments, environment variables and auxv passed to the app.
```
ustack_top -＞ ustack_pointer
ustack_bottom -> ustack_start

equation314 commented on src/loader.rs:
```diff
@@ -145,14 +140,22 @@ pub(crate) fn load_user_app(name: &str) -> ELFInfo {
+
+    let elf_offset = kernel_elf_parser::get_elf_base_addr(&elf, base_addr.as_usize()).unwrap();
+    assert!(
+        elf_offset & 0xfff == 0,
```
Use `memory_addr::is_aligned_4k`

equation314 commented on src/syscall_imp/fs/mod.rs:
```diff
@@ -0,0 +1,5 @@
+mod ctl;
+mod io;
+
+pub(crate) use ctl::*;
```
`self::ctl::*`

equation314 commented on src/syscall_imp/mm/mmap.rs:
```diff
@@ -0,0 +1,100 @@
+
+        aspace
+            .map_alloc(start_addr, length, permission_flags.into(), false)
+            .map_err(<AxError as From<_>>::from)?;
```
Is this line necessary? Can we just use `?` at the previous line?

equation314 commented on src/syscall_imp/mod.rs:
```diff
@@ -0,0 +1,66 @@
+use syscalls::Sysno;
+use task::*;
+use time::*;
+/// Macro to generate syscall body
```
Blank line

equation314 commented on src/syscall_imp/task/thread.rs:
```diff
@@ -0,0 +1,96 @@
+///
+/// It is only avaliable on x86_64, and is not convenient
+/// to generate automatically via c_to_rust binding.
+enum ArchPrctlCode {
```
Move comment to the top, with attributes followed.

equation314 commented on src/syscall_imp/task/thread.rs:
```diff
@@ -0,0 +1,96 @@
+            // TODO: check the legality of the address
+            Ok(ArchPrctlCode::SetFs) => {
+                unsafe {
+                    axhal::arch::write_thread_pointer(*addr);
```
For “set” operation, it should use `addr` directly without dereference. Please refer to the specification.

equation314 commented on src/syscall_imp/task/mod.rs:
```diff
@@ -0,0 +1,4 @@
+mod schedule;
+mod thread;
+pub(crate) use schedule::*;
```
Blank line

equation314 commented on src/main.rs:
```diff
@@ -1,23 +1,22 @@
 extern crate alloc;
 extern crate axstd;
-
+#[allow(unused)]
```
Move it into config.rs

Azure-stars replied:
ok

equation314 commented on configs/aarch64.toml:
```diff
@@ -0,0 +1,17 @@
+# The base address of the user space.
+user_space_base = 0x1000
```
It not used? We still create `uspace` with the parameters hardcoded in `axmm`

Azure-stars replied:
It will be used in the next PR, but here i will remove it.

Azure-stars committed:
[style] Fix code style according to the PR content

equation314 commented on Cargo.toml:
```diff
@@ -10,14 +10,24 @@ repository = "https://github.com/arceos-org/starry-next"
 
-axstd = { git = "https://github.com/arceos-org/arceos.git", features = ["paging"] }
-arceos_posix_api = { git = "https://github.com/arceos-org/arceos.git" }
+kernel-elf-parser = "0.1.0"
```
Remove the blank line before

Azure-stars committed:
[fix] Set the base address for static-ELF as 0 by default

Azure-stars committed:
[feat] Support complie and test musl app on rv64 and aarch64

Azure-stars committed:
[config] Fix the format for config file

equation314 commented on apps/libc/Makefile:
```diff
@@ -0,0 +1,73 @@
+		for app in $(shell find rust -name Cargo.toml); do \
+			echo "Building $$(dirname $${app})"; \
+			app_name=$$(basename $$(dirname $${app})); \
+			RUSTFLAGS=$(RUSTFLAGS) cargo build --release --target $(RUST_TARGET) --manifest-path $${app} ; \
```
Add `export RUSTFLAGS` before and remove `RUSTFLAGS=$(RUSTFLAGS)`

equation314 commented on apps/nimbos/c/lib/syscall_ids.h.no_x86.in:
Move to `nimbos/c/lib/arch/aarch64` and `nimbos/c/lib/arch/riscv`

equation314 commented on debug.S:
Remove this file

equation314 commented on src/syscall_imp/mm/mmap.rs:
```diff
@@ -0,0 +1,98 @@
+    /// permissions for sys_mmap
+    ///
+    /// See <https://github.com/bminor/glibc/blob/master/bits/mman.h>
+    struct MmapProt: i32 {
```
Put `#[derive(Debug)]`  between comments and the struct.

equation314 commented on src/syscall_imp/mm/mmap.rs:
```diff
@@ -0,0 +1,98 @@
+    /// flags for sys_mmap
+    ///
+    /// See <https://github.com/bminor/glibc/blob/master/bits/mman.h>
+    struct MmapFlags: i32 {
```
See above

equation314 commented on build.rs:
```diff
@@ -70,3 +73,45 @@ _app_count:
+            let key_name = key.to_uppercase().replace('-', "_");
+            match value {
+                toml_edit::Value::Integer(i) => {
+                    writeln!(f, "pub(crate) const {}: usize = {};", key_name, i)?;
```
`pub` is OK, see constants in axconfig. 

You may need to add `#![allow(dead_code)]` in the generated code.

equation314 commented on src/syscall_imp/mod.rs:
```diff
@@ -0,0 +1,67 @@
+use syscalls::Sysno;
+use task::*;
+use time::*;
+
```
Importing style in ArceOS:

```rs
mod sub_mod1;
mod sub_mod2;

use core::foo::Bar;
use alloc::bar::Foo;

use ext_crate::bar;

use crate::baz;
use self::sub_mod1::foo;
use super::sup_mod::qux;

pub use ...
```

Azure-stars committed:
[style] format the code style

Azure-stars committed:
[style] format crate import style:
```rust
mod sub_mod1;
mod sub_mod2;

use core::foo::Bar;
use alloc::bar::Foo;

use ext_crate::bar;

use crate::baz;
use self::sub_mod1::foo;
use super::sup_mod::qux;

pub use ...
```

equation314 commented on apps/nimbos/c/CMakeLists.txt:
```diff
@@ -60,9 +60,19 @@ foreach(PATH ${SRCS})
+    add_custom_command(
+        OUTPUT syscall_ids.h
+        COMMAND sed ARGS -n -e s/__NR_/SYS_/p
+            < ${CMAKE_SOURCE_DIR}/lib/syscall_ids.h.no_x86.in
```
`${ARCH}/syscall_ids.h.in`

Azure-stars replied:
Done

equation314 commented on apps/nimbos/c/lib/syscall_ids.h.no_x86.in:
It's better to put it into a separate directory for each arch.

Azure-stars replied:
Put it into /lib/arch directory

equation314 commented on build.rs:
```diff
@@ -70,3 +73,57 @@ _app_count:
+                    writeln!(
+                        f,
+                        "
+                    #[allow(dead_code)]\n
```
Put `#![allow(dead_code)]` at the first line and remove this line.

equation314 commented on src/main.rs:
```diff
@@ -26,6 +26,7 @@ fn main() {
         .split(',')
         .filter(|&x| !x.is_empty());
     for testcase in testcases {
+        log::info!("Running testcase: {}", testcase);
```
`log::info!` => `info!`

equation314 commented on apps/libc/Makefile:
```diff
@@ -0,0 +1,74 @@
+		for app in $(shell find rust -name Cargo.toml); do \
+			echo "Building $$(dirname $${app})"; \
+			app_name=$$(basename $$(dirname $${app})); \
+      export RUSTFLAGS=$(RUSTFLAGS); \
```
Put `export RUSTFLAGS=$(RUSTFLAGS)` outside the `build_rust` block so that you don't have to write it next time when run `cargo build`.

Azure-stars committed:
[style] format code style

equation314 commented on apps/nimbos/c/CMakeLists.txt:
```diff
@@ -60,9 +60,25 @@ foreach(PATH ${SRCS})
     )
 endforeach()
 
+# If arch is not x86_64, we need to use different syscall ids
```
Remove comments

equation314 commented on build.rs:
```diff
@@ -70,3 +73,46 @@ _app_count:
+        writeln!(f, "{}", comments)?;
+        if let Item::Value(value) = item {
+            let key_name = key.to_uppercase().replace('-', "_");
+            writeln!(f, "#[allow(dead_code)]")?;
```
`#[allow(dead_code)]` -> `#![allow(dead_code)]`, and move  to the line between 97 and 98.
